sys/net/routing/ng_rpl/ng_rpl_control_messages.c

```diff
@@ -0,0 +1,617 @@
+    transit->e_flags = 0;
+    transit->path_control = 0;
+    transit->path_sequence = 0;
+    transit->path_lifetime = lifetime;
+
+    _ng_rpl_send(pkt, NULL, destination, &dodag->dodag_id);
+
+    NG_RPL_COUNTER_INCREMENT(dodag->dao_seq);
+}
+
+void ng_rpl_send_DAO_ACK(ng_rpl_dodag_t *dodag, ng_ipv6_addr_t *destination, uint8_t seq)
+{
+    if (dodag == NULL) {
+        DEBUG("RPL: Error - trying to send DAO-ACK without being part of a dodag.\n");
+        return;
+    }
+
+    ng_pktsnip_t *pkt;
+    ng_icmpv6_hdr_t *icmp;
+    ng_rpl_dao_ack_t *dao_ack;
+    int size = sizeof(ng_icmpv6_hdr_t) + sizeof(ng_rpl_dao_ack_t);
+
+    if ((pkt = ng_icmpv6_build(NULL, NG_ICMPV6_RPL_CTRL, NG_RPL_ICMPV6_CODE_DAO_ACK, size)) == NULL) {
+        DEBUG("RPL: Send DAOACK - no space left in packet buffer\n");
+        return;
+    }
+
+    icmp = (ng_icmpv6_hdr_t *)pkt->data;
+    dao_ack = (ng_rpl_dao_ack_t *)(icmp + 1);
+
+    dao_ack->instance_id = dodag->instance->id;
+    /* set the D flag to indicate that a DODAG id is present */
+    dao_ack->d_reserved = (1 << 7);
+    dao_ack->dao_sequence = seq;
+    dao_ack->status = 0;
+    dao_ack->dodag_id = dodag->dodag_id;
+
+    _ng_rpl_send(pkt, NULL, destination, &dodag->dodag_id);
+}
+
+void ng_rpl_recv_DAO(ng_rpl_dao_t *dao, ng_ipv6_addr_t *src, uint16_t len)
+{
+    ng_rpl_instance_t *inst = NULL;
+    ng_rpl_dodag_t *dodag = NULL;
+    if ((inst = ng_rpl_instance_get(dao->instance_id)) == NULL) {
+        DEBUG("RPL: DAO with unknown instance id (%d) received\n", dao->instance_id);
+        return;
+    }
+
+    /* check if the D flag is set before accessing the DODAG id */
+    if (!(dao->k_d_flags & (1 << 6))) {
+        DEBUG("RPL: DAO with D flag unset - global instances not supported\n");
+        return;
+    }
+
+    if ((dodag = ng_rpl_dodag_get(inst, &dao->dodag_id)) == NULL) {
+        DEBUG("RPL: DAO with unknown DODAG id (%s)\n", ng_ipv6_addr_to_str(addr_str,
+                    &dao->dodag_id, sizeof(addr_str)));
+        return;
+    }
+
+    len -= (sizeof(ng_rpl_dao_t) + sizeof(ng_icmpv6_hdr_t));
+    if(!_parse_options(NG_RPL_ICMPV6_CODE_DAO, dodag, (ng_rpl_opt_t *) (dao + 1), len, src)) {
+        ng_rpl_dodag_remove(dodag);
+        return;
+    }
+
+    /* send a DAO-ACK if K flag is set */
+    if (dao->k_d_flags & (1 << 7)) {
+        ng_rpl_send_DAO_ACK(dodag, src, dao->dao_sequence);
+    }
+
+    ng_rpl_delay_dao(dodag);
+}
+
+void ng_rpl_recv_DAO_ACK(ng_rpl_dao_ack_t *dao_ack)
+{
+    ng_rpl_instance_t *inst = NULL;
+    ng_rpl_dodag_t *dodag = NULL;
+    if ((inst = ng_rpl_instance_get(dao_ack->instance_id)) == NULL) {
+        DEBUG("RPL: DAO-ACK with unknown instance id (%d) received\n", dao_ack->instance_id);
+        return;
+    }
+
+    /* check if the D flag is set before accessing the DODAG id */
+    if (!(dao_ack->d_reserved & (1 << 7))) {
+        DEBUG("RPL: DAO-ACK with D flag unset - global instances not supported\n");
```

Comment:
In a follow up: update `ng_rpl_dodag_get` to handle NULL as the second parameter and return the first DODAG in the instance. Bonus points: check that there is exactly one DODAG assigned to the instance.

Contiki sends DAO ACK without DODAG ID, at least if the RPLInstanceID is within the global range (<127 IIRC)